src/assign.c

```diff
@@ -152,13 +152,21 @@ static SEXP shallow(SEXP dt, SEXP cols, R_len_t n)
   R_len_t i,l;
   int protecti=0;
   SEXP newdt = PROTECT(allocVector(VECSXP, n)); protecti++;   // to do, use growVector here?
-  //copyMostAttrib(dt, newdt);   // including class
-  DUPLICATE_ATTRIB(newdt, dt);
+  SHALLOW_DUPLICATE_ATTRIB(newdt, dt);
   // TO DO: keepattr() would be faster, but can't because shallow isn't merely a shallow copy. It
   //        also increases truelength. Perhaps make that distinction, then, and split out, but marked
   //        so that the next change knows to duplicate.
-  //        Does copyMostAttrib duplicate each attrib or does it point? It seems to point, hence DUPLICATE_ATTRIB
-  //        for now otherwise example(merge.data.table) fails (since attr(d4,"sorted") gets written by setnames).
+  //        keepattr() also merely points to the entire attrbutes list and thus doesn't allow replacing
+  //        some of its elements.
+  
+  // We copy all attributes that refer to column names so that calling setnames on either
+  // the original or the shallow copy doesn't break anything.
+  SEXP index = PROTECT(getAttrib(dt, sym_index)); protecti++;
+  setAttrib(newdt, sym_index, shallow_duplicate(index));
```

Comment:
`shallow_duplicate` was AFAIK designed to be used on `VECSXP`, `LISTSXP`, etc. You call it here on `INTSXP`. It would be good to figure out proper/better? way to shallow copy index. PS. I do exactly the same in #4386.
Hope @mattdowle will have good idea how to handle that. Maybe we need to change index structure, and wrap it into extra list?

Reply:
Looking at https://github.com/wch/r-source/blob/trunk/src/main/duplicate.c that doesn't seem to be an issue. It starts out exactly like `duplicate` in that it makes a copy of the `INTSXP` and then a shallow copy of the attributes. And the flow looks the same for R 3.1 as far as I can tell.
Initially, I had done this more manually by allocating a zero length `INTSXP` and making a shallow copy of the attributes but this seemed more succinct.

Reply:
I made a dedicated issue to have it well discussed: https://github.com/Rdatatable/data.table/issues/4467

Reply:
This `INTSXP` being copied here is an empty one though (the dummy object upon which the indexes are attached as attributes). So I'm not sure what's improper here in the PR as it stands now. @jangorecki Do you mean that it's possible that `shallow_duplicate()` might not work like this in future in R on this case of being passed INTSXP? If so, that's a good spot but on balance it feels right to use it here for this. And the tests in CI will spot if and when R-devel makes any change to `shallow_duplicate()`.

Reply:
Thanks Matt. Good spot. Sorry for confusion. Indeed copy of `integer()` (a placeholder to which we are attaching indices) is not an issue here. The issue is copying of indices itself, which is not happening here. So the related items I was linking here are not really matching use case here.